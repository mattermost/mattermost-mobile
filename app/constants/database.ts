// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import keyMirror from '@utils/key_mirror';

export const MM_TABLES = {
    APP: {
        INFO: 'Info',
        GLOBAL: 'Global',
        SERVERS: 'Servers',
    },
    SERVER: {
        CATEGORY: 'Category',
        CATEGORY_CHANNEL: 'CategoryChannel',
        CHANNEL: 'Channel',
        CHANNEL_BOOKMARK: 'ChannelBookmark',
        CHANNEL_INFO: 'ChannelInfo',
        CHANNEL_MEMBERSHIP: 'ChannelMembership',
        CONFIG: 'Config',
        CUSTOM_EMOJI: 'CustomEmoji',
        DRAFT: 'Draft',
        FILE: 'File',
        GROUP: 'Group',
        GROUP_CHANNEL: 'GroupChannel',
        GROUP_MEMBERSHIP: 'GroupMembership',
        GROUP_TEAM: 'GroupTeam',
        MY_CHANNEL: 'MyChannel',
        MY_CHANNEL_SETTINGS: 'MyChannelSettings',
        MY_TEAM: 'MyTeam',
        POST: 'Post',
        POSTS_IN_CHANNEL: 'PostsInChannel',
        POSTS_IN_THREAD: 'PostsInThread',
        PREFERENCE: 'Preference',
        REACTION: 'Reaction',
        ROLE: 'Role',
        SCHEDULED_POST: 'ScheduledPost',
        SYSTEM: 'System',
        TEAM: 'Team',
        TEAM_CHANNEL_HISTORY: 'TeamChannelHistory',
        TEAM_MEMBERSHIP: 'TeamMembership',
        TEAM_SEARCH_HISTORY: 'TeamSearchHistory',
        THREAD: 'Thread',
        THREADS_IN_TEAM: 'ThreadsInTeam',
        THREAD_PARTICIPANT: 'ThreadParticipant',
        TEAM_THREADS_SYNC: 'TeamThreadsSync',
        USER: 'User',
    },
};

export const MIGRATION_EVENTS = keyMirror({
    MIGRATION_ERROR: null,
    MIGRATION_STARTED: null,
    MIGRATION_SUCCESS: null,
});

export const SYSTEM_IDENTIFIERS = {
    CURRENT_CHANNEL_ID: 'currentChannelId',
    LAST_UNREAD_CHANNEL_ID: 'lastUnreadChannelId',
    CURRENT_TEAM_ID: 'currentTeamId',
    CURRENT_USER_ID: 'currentUserId',
    DATA_RETENTION_POLICIES: 'dataRetentionPolicies',
    EXPANDED_LINKS: 'expandedLinks',
    GRANULAR_DATA_RETENTION_POLICIES: 'granularDataRetentionPolicies',
    LAST_DATA_RETENTION_RUN: 'lastDataRetentionRun',
    GLOBAL_THREADS_TAB: 'globalThreadsTab',
    LAST_DISMISSED_BANNER: 'lastDismissedBanner',
    LAST_SERVER_VERSION_CHECK: 'LastServerVersionCheck',
    LICENSE: 'license',
    ONLY_UNREADS: 'onlyUnreads',
    PUSH_VERIFICATION_STATUS: 'pushVerificationStatus',
    RECENT_CUSTOM_STATUS: 'recentCustomStatus',
    RECENT_MENTIONS: 'recentMentions',
    RECENT_REACTIONS: 'recentReactions',
    SESSION_EXPIRATION: 'sessionExpiration',
    TEAM_HISTORY: 'teamHistory',
    WEBSOCKET: 'WebSocket',
};

export const GLOBAL_IDENTIFIERS = {
    DEVICE_TOKEN: 'deviceToken',
    DONT_ASK_FOR_REVIEW: 'dontAskForReview',
    FIRST_LAUNCH: 'firstLaunch',
    LAST_ASK_FOR_REVIEW: 'lastAskForReview',
    ONBOARDING: 'onboarding',
    LAST_VIEWED_CHANNEL: 'lastViewedChannel',
    LAST_VIEWED_THREAD: 'lastViewedThread',
    PUSH_DISABLED_ACK: 'pushDisabledAck',
};

export enum OperationType {
    CREATE = 'CREATE',
    UPDATE = 'UPDATE',
    DELETE = 'DELETE',
  }

// The only two types of databases in the app
export enum DatabaseType {
    DEFAULT,
    SERVER,
  }

export default {
    GLOBAL_IDENTIFIERS,
    MM_TABLES,
    MIGRATION_EVENTS,
    SYSTEM_IDENTIFIERS,
};
