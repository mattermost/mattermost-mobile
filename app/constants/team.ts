// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export const ALL_TEAMS_ID = '';

export default {
    ALL_TEAMS_ID,
};
