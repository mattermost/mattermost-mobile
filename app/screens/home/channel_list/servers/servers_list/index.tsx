// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {BottomSheetFlatList} from '@gorhom/bottom-sheet';
import React, {useCallback, useMemo} from 'react';
import {useIntl} from 'react-intl';
import {FlatList, StyleSheet, View, type ListRenderItemInfo} from 'react-native';

import {useServerUrl} from '@context/server';
import {useTheme} from '@context/theme';
import {useBottomSheetListsFix} from '@hooks/bottom_sheet_lists_fix';
import {useIsTablet} from '@hooks/device';
import {BUTTON_HEIGHT} from '@screens/bottom_sheet';
import BottomSheetContent from '@screens/bottom_sheet/content';
import {addNewServer} from '@utils/server';

import ServerItem from './server_item';

import type ServersModel from '@typings/database/models/app/servers';

export {default as AddServerButton} from './add_server_button';

type Props = {
    servers: ServersModel[];
}

const styles = StyleSheet.create({
    container: {
        flexGrow: 1,
    },
    contentContainer: {
        marginVertical: 4,
    },
    serverList: {
        marginBottom: BUTTON_HEIGHT,
    },
});

const keyExtractor = (item: ServersModel) => item.url;

const ServerList = ({servers}: Props) => {
    const intl = useIntl();
    const isTablet = useIsTablet();
    const serverUrl = useServerUrl();
    const theme = useTheme();
    const {enabled, panResponder} = useBottomSheetListsFix();

    const onAddServer = useCallback(async () => {
        addNewServer(theme);
    }, [servers]);

    const renderServer = useCallback(({item: t, index}: ListRenderItemInfo<ServersModel>) => {
        return (
            <ServerItem
                highlight={index === 0}
                isActive={t.url === serverUrl}
                server={t}
            />
        );
    }, []);

    const List = useMemo(() => (isTablet ? FlatList : BottomSheetFlatList), [isTablet]);

    return (
        <BottomSheetContent
            buttonIcon='plus'
            buttonText={intl.formatMessage({id: 'servers.create_button', defaultMessage: 'Add a server'})}
            onPress={onAddServer}
            showButton={isTablet}
            showTitle={!isTablet}
            testID='server_list'
            title={intl.formatMessage({id: 'your.servers', defaultMessage: 'Your servers'})}
        >
            <View style={[styles.container, {marginTop: isTablet ? 12 : 0}]}>
                <List
                    data={servers}
                    style={styles.serverList}
                    renderItem={renderServer}
                    keyExtractor={keyExtractor}
                    contentContainerStyle={styles.contentContainer}
                    scrollEnabled={enabled}
                    {...panResponder.panHandlers}
                />
            </View>
        </BottomSheetContent>
    );
};

export default ServerList;
