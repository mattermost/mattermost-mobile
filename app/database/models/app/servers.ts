// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Model} from '@nozbe/watermelondb';
import {field} from '@nozbe/watermelondb/decorators';

import {MM_TABLES} from '@constants/database';

import type ServersModelInterface from '@typings/database/models/app/servers';

const {SERVERS} = MM_TABLES.APP;

/**
 * The Server model will help us to identify the various servers a user will log in; in the context of
 * multi-server support system.  The db_path field will hold the App-Groups file-path
 */
export default class ServersModel extends Model implements ServersModelInterface {
    /** table (name) : servers */
    static table = SERVERS;

    /** db_path : The file path where the database is stored */
    @field('db_path') dbPath!: string;

    /** display_name : The server display name */
    @field('display_name') displayName!: string;

    /** url : The online address for the Mattermost server */
    @field('url') url!: string;

    /** last_active_at: The last time this server was active */
    @field('last_active_at') lastActiveAt!: number;

    /** identifier: Determines the installation identifier of a server */
    @field('identifier') identifier!: string;
}
