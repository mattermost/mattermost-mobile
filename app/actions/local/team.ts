// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DatabaseManager from '@database/manager';
import {prepareDeleteTeam, getMyTeamById, queryTeamSearchHistoryByTeamId, removeTeamFromTeamHistory, getTeamSearchHistoryById, getTeamById} from '@queries/servers/team';
import {logError} from '@utils/log';

import type Model from '@nozbe/watermelondb/Model';

export const MAX_TEAM_SEARCHES = 15;

export async function removeUserFromTeam(serverUrl: string, teamId: string) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const myTeam = await getMyTeamById(database, teamId);
        if (myTeam) {
            const team = await getTeamById(database, myTeam.id);
            if (!team) {
                throw new Error('Team not found');
            }
            const models = await prepareDeleteTeam(serverUrl, team);
            const system = await removeTeamFromTeamHistory(operator, team.id, true);
            if (system) {
                models.push(...system);
            }
            if (models.length) {
                await operator.batchRecords(models, 'removeUserFromTeam');
            }
        }

        return {error: undefined};
    } catch (error) {
        logError('Failed removeUserFromTeam', error);
        return {error};
    }
}

export async function addSearchToTeamSearchHistory(serverUrl: string, teamId: string, terms: string) {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const newSearch: TeamSearchHistory = {
            created_at: Date.now(),
            display_term: terms,
            term: terms,
            team_id: teamId,
        };

        const models: Model[] = [];
        const searchModels = await operator.handleTeamSearchHistory({teamSearchHistories: [newSearch], prepareRecordsOnly: true});
        const searchModel = searchModels[0];

        models.push(searchModel);

        // determine if need to delete the oldest entry
        if (searchModel._raw._changed !== 'created_at') {
            const teamSearchHistory = await queryTeamSearchHistoryByTeamId(database, teamId).fetch();
            if (teamSearchHistory.length > MAX_TEAM_SEARCHES) {
                const lastSearches = teamSearchHistory.slice(MAX_TEAM_SEARCHES);
                for (const lastSearch of lastSearches) {
                    models.push(lastSearch.prepareDestroyPermanently());
                }
            }
        }

        await operator.batchRecords(models, 'addSearchToTeamHistory');
        return {searchModel};
    } catch (error) {
        logError('Failed addSearchToTeamSearchHistory', error);
        return {error};
    }
}

export async function removeSearchFromTeamSearchHistory(serverUrl: string, id: string) {
    try {
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const teamSearch = await getTeamSearchHistoryById(database, id);
        if (teamSearch) {
            await database.write(async () => {
                await teamSearch.destroyPermanently();
            });
        }
        return {teamSearch};
    } catch (error) {
        logError('Failed removeSearchFromTeamSearchHistory', error);
        return {error};
    }
}

