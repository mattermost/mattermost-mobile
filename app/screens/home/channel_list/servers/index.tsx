// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useCallback, useEffect, useImperativeHandle, useRef, useState} from 'react';
import {useIntl} from 'react-intl';
import {Dimensions, StyleSheet} from 'react-native';

import ServerIcon from '@components/server_icon';
import {useServerUrl} from '@context/server';
import {useTheme} from '@context/theme';
import {subscribeAllServers} from '@database/subscription/servers';
import {subscribeUnreadAndMentionsByServer, type UnreadObserverArgs} from '@database/subscription/unreads';
import {useIsTablet} from '@hooks/device';
import {BUTTON_HEIGHT, TITLE_HEIGHT} from '@screens/bottom_sheet';
import {bottomSheet} from '@screens/navigation';
import {bottomSheetSnapPoint} from '@utils/helpers';
import {sortServersByDisplayName} from '@utils/server';

import ServerList, {AddServerButton} from './servers_list';

import type ServersModel from '@typings/database/models/app/servers';
import type {UnreadMessages, UnreadSubscription} from '@typings/database/subscriptions';

export type ServersRef = {
    openServers: () => void;
}

export const SERVER_ITEM_HEIGHT = 75;
export const PUSH_ALERT_TEXT_HEIGHT = 42;
const subscriptions: Map<string, UnreadSubscription> = new Map();

const styles = StyleSheet.create({
    icon: {
        alignItems: 'center',
        justifyContent: 'center',
        position: 'absolute',
        zIndex: 10,
        top: 10,
        left: 16,
        width: 40,
        height: 40,
    },
});

const Servers = React.forwardRef<ServersRef>((_, ref) => {
    const intl = useIntl();
    const [total, setTotal] = useState<UnreadMessages>({mentions: 0, unread: false});
    const registeredServers = useRef<ServersModel[]|undefined>();
    const currentServerUrl = useServerUrl();
    const isTablet = useIsTablet();
    const theme = useTheme();

    const updateTotal = () => {
        let unread = false;
        let mentions = 0;
        subscriptions.forEach((value) => {
            unread = unread || value.unread;
            mentions += value.mentions;
        });
        setTotal({mentions, unread});
    };

    const unreadsSubscription = (serverUrl: string, {myChannels, settings, threadMentionCount, threadUnreads}: UnreadObserverArgs) => {
        const unreads = subscriptions.get(serverUrl);
        if (unreads) {
            let mentions = 0;
            let unread = Boolean(threadUnreads);
            for (const myChannel of myChannels) {
                const isMuted = settings?.[myChannel.id]?.mark_unread === 'mention';
                mentions += isMuted ? 0 : myChannel.mentionsCount;
                unread = unread || (myChannel.isUnread && !isMuted);
            }

            unreads.mentions = mentions + threadMentionCount;
            unreads.unread = unread;
            subscriptions.set(serverUrl, unreads);
            updateTotal();
        }
    };

    const serversObserver = async (servers: ServersModel[]) => {
        registeredServers.current = sortServersByDisplayName(servers, intl);

        // unsubscribe mentions from servers that were removed
        const allUrls = new Set(servers.map((s) => s.url));
        const subscriptionsToRemove = [...subscriptions].filter(([key]) => !allUrls.has(key));
        for (const [key, map] of subscriptionsToRemove) {
            map.subscription?.unsubscribe();
            subscriptions.delete(key);
            updateTotal();
        }

        for (const server of servers) {
            const {lastActiveAt, url} = server;
            if (lastActiveAt && (url !== currentServerUrl) && !subscriptions.has(url)) {
                const unreads: UnreadSubscription = {
                    mentions: 0,
                    unread: false,
                };
                subscriptions.set(url, unreads);
                unreads.subscription = subscribeUnreadAndMentionsByServer(url, unreadsSubscription);
            } else if ((!lastActiveAt || (url === currentServerUrl)) && subscriptions.has(url)) {
                subscriptions.get(url)?.subscription?.unsubscribe();
                subscriptions.delete(url);
                updateTotal();
            }
        }
    };

    const onPress = useCallback(() => {
        if (registeredServers.current?.length) {
            const renderContent = () => {
                return (
                    <ServerList servers={registeredServers.current!}/>
                );
            };
            const maxScreenHeight = Math.ceil(0.6 * Dimensions.get('window').height);
            const maxSnapPoint = Math.min(
                maxScreenHeight,
                bottomSheetSnapPoint(registeredServers.current.length, SERVER_ITEM_HEIGHT) + TITLE_HEIGHT + BUTTON_HEIGHT +
                    (registeredServers.current.filter((s: ServersModel) => s.lastActiveAt).length * PUSH_ALERT_TEXT_HEIGHT),
            );

            const snapPoints: Array<string | number> = [
                1,
                maxSnapPoint,
            ];
            if (maxSnapPoint === maxScreenHeight) {
                snapPoints.push('80%');
            }

            const closeButtonId = 'close-your-servers';
            bottomSheet({
                closeButtonId,
                renderContent,
                footerComponent: isTablet ? undefined : AddServerButton,
                snapPoints,
                theme,
                title: intl.formatMessage({id: 'your.servers', defaultMessage: 'Your servers'}),
            });
        }
    }, [intl, isTablet, theme]);

    useImperativeHandle(ref, () => ({
        openServers: onPress,
    }), [onPress]);

    useEffect(() => {
        const subscription = subscribeAllServers(serversObserver);

        return () => {
            subscription?.unsubscribe();
            subscriptions.forEach((unreads) => {
                unreads.subscription?.unsubscribe();
            });
            subscriptions.clear();
        };
    }, []);

    return (
        <ServerIcon
            hasUnreads={total.unread}
            mentionCount={total.mentions}
            onPress={onPress}
            style={styles.icon}
            testID={'channel_list.servers.server_icon'}
            badgeBorderColor={theme.sidebarBg}
            badgeBackgroundColor={theme.mentionBg}
            badgeColor={theme.mentionColor}
        />
    );
});

Servers.displayName = 'Servers';

export default Servers;
