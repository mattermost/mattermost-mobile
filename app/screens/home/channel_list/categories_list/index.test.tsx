// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {act} from '@testing-library/react-native';
import React from 'react';

import {SYSTEM_IDENTIFIERS} from '@constants/database';
import {getTeamById} from '@queries/servers/team';
import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import CategoriesList from './categories_list';

import type ServerDataOperator from '@database/operator/server_data_operator';
import type Database from '@nozbe/watermelondb/Database';

describe('components/categories_list', () => {
    let database: Database;
    let operator: ServerDataOperator;
    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
        operator = server.operator;

        const team = await getTeamById(database, TestHelper.basicTeam!.id);
        await database.write(async () => {
            await team?.update(() => {
                team.displayName = 'Test Team!';
            });
        });
    });

    it('should render', () => {
        const wrapper = renderWithEverything(
            <CategoriesList
                moreThanOneTeam={false}
                hasChannels={true}
                draftsCount={0}
            />,
            {database},
        );
        expect(wrapper.toJSON()).toBeTruthy();
    });

    it('should render channel list with thread menu', () => {
        jest.useFakeTimers();
        const wrapper = renderWithEverything(
            <CategoriesList
                isCRTEnabled={true}
                moreThanOneTeam={false}
                hasChannels={true}
                draftsCount={0}
            />,
            {database},
        );
        act(() => {
            jest.runAllTimers();
        });
        expect(wrapper.toJSON()).toBeTruthy();
        jest.useRealTimers();
    });

    it('should render channel list with Draft menu', () => {
        const wrapper = renderWithEverything(
            <CategoriesList
                isCRTEnabled={true}
                moreThanOneTeam={false}
                hasChannels={true}
                draftsCount={1}
            />,
            {database},
        );
        expect(wrapper.getByText('Drafts')).toBeTruthy();
    });

    it('should render team error', async () => {
        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: ''}],
            prepareRecordsOnly: false,
        });

        jest.useFakeTimers();
        const wrapper = renderWithEverything(
            <CategoriesList
                moreThanOneTeam={false}
                hasChannels={true}
                draftsCount={0}
            />,
            {database},
        );

        act(() => {
            jest.runAllTimers();
        });
        expect(wrapper.toJSON()).toMatchSnapshot();
        jest.useRealTimers();

        await operator.handleSystem({
            systems: [{id: SYSTEM_IDENTIFIERS.CURRENT_TEAM_ID, value: TestHelper.basicTeam!.id}],
            prepareRecordsOnly: false,
        });
    });

    it('should render channels error', () => {
        jest.useFakeTimers();
        const wrapper = renderWithEverything(
            <CategoriesList
                moreThanOneTeam={true}
                hasChannels={false}
                draftsCount={0}
            />,
            {database},
        );
        act(() => {
            jest.runAllTimers();
        });
        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
