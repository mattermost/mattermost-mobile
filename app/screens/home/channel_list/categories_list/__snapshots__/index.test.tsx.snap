// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/categories_list should render channels error 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#1e325c",
        "flex": 1,
        "paddingTop": 10,
      },
      {
        "maxWidth": 750,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "marginLeft": 0,
        },
        {
          "paddingLeft": 18,
          "paddingRight": 20,
        },
      ]
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          {
            "width": "85%",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessible={true}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text
              ellipsizeMode="tail"
              numberOfLines={2}
              style={
                {
                  "color": "#ffffff",
                  "fontFamily": "Metropolis-SemiBold",
                  "fontSize": 28,
                  "fontWeight": "600",
                  "lineHeight": 36,
                }
              }
              testID="channel_list_header.team_display_name"
            >
              Test Team!
            </Text>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingRight": 60,
            }
          }
        >
          <Text
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              {
                "color": "rgba(255,255,255,0.64)",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 11,
                "fontWeight": "600",
                "lineHeight": 16,
              }
            }
            testID="channel_list_header.server_display_name"
          />
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          {
            "bottom": 30,
            "left": 20,
            "right": 20,
            "top": 10,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(255,255,255,0.08)",
            "borderRadius": 14,
            "height": 28,
            "justifyContent": "center",
            "marginTop": 7,
            "opacity": 1,
            "width": 28,
          }
        }
        testID="channel_list_header.plus.button"
      >
        <Icon
          name="plus"
          style={
            {
              "color": "rgba(255,255,255,0.8)",
              "fontSize": 18,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load Test Team!
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this team.
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "marginTop": 24,
          "opacity": 1,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            {
              "color": "#1c58d9",
            },
          ]
        }
      >
        Retry
      </Text>
    </View>
  </View>
</View>
`;

exports[`components/categories_list should render team error 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#1e325c",
        "flex": 1,
        "paddingTop": 10,
      },
      {
        "maxWidth": 750,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "marginLeft": 0,
        },
        {
          "paddingLeft": 18,
          "paddingRight": 20,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "space-between",
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "space-between",
            },
            {
              "flex": 1,
            },
          ]
        }
      >
        <Text
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            {
              "color": "rgba(255,255,255,0.64)",
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 20,
            }
          }
          testID="channel_list_header.team_display_name"
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="channel_list_header.logout.button"
      >
        <Text
          style={
            {
              "color": "rgba(255,255,255,0.64)",
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 20,
            }
          }
          testID="channel_list_header.team_display_name"
        >
          Log out
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "paddingLeft": 18,
        "paddingRight": 20,
      }
    }
  >
    <View
      style={
        {
          "marginRight": 8,
          "opacity": 1,
          "width": 40,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "backgroundColor": "rgba(255,255,255,0.12)",
                "borderRadius": 8,
                "height": 40,
                "justifyContent": "center",
                "marginVertical": 20,
                "width": 40,
              },
              false,
            ]
          }
        >
          <Icon
            color="rgba(255,255,255,0.56)"
            name="filter-variant"
            size={24}
          />
        </View>
      </View>
    </View>
    <View
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "rgba(255,255,255,0.12)",
          "borderRadius": 8,
          "flex": 1,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "flex-start",
          "marginVertical": 20,
          "padding": 8,
        }
      }
      testID="channel_list_subheader.search_field.button"
    >
      <Icon
        name="magnify"
        style={
          {
            "color": "rgba(255,255,255,0.72)",
            "fontSize": 24,
            "width": 24,
          }
        }
      />
      <Text
        style={
          {
            "color": "rgba(255,255,255,0.72)",
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
            "marginLeft": 5,
            "marginTop": 1,
          }
        }
      >
        Find channels...
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(255,255,255,0.08)",
          "borderRadius": 60,
          "height": 120,
          "justifyContent": "center",
          "width": 120,
        }
      }
    >
      <Icon
        name="alert-circle-outline"
        style={
          {
            "color": "rgba(255,255,255,0.48)",
            "fontSize": 72,
            "lineHeight": 72,
          }
        }
      />
    </View>
    <Text
      style={
        [
          {
            "fontFamily": "Metropolis-SemiBold",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 28,
          },
          {
            "color": "#ffffff",
            "marginTop": 20,
            "textAlign": "center",
          },
        ]
      }
    >
      Couldn't load categories for 
    </Text>
    <Text
      style={
        [
          {
            "fontFamily": "OpenSans",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 24,
          },
          {
            "color": "#ffffff",
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      There was a problem loading content for this server.
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 4,
          "marginTop": 24,
          "opacity": 1,
          "paddingHorizontal": 24,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            {
              "color": "#1c58d9",
            },
          ]
        }
      >
        Retry
      </Text>
    </View>
  </View>
</View>
`;
